quinn-proto/src/connection/bytes_source.rs

```diff
@@ -0,0 +1,249 @@
+//! Defines types which can be lazily converted into `Bytes` chunks
+
+use bytes::Bytes;
+
+/// A source of one or more buffers which can be converted into `Bytes` buffers on demand.
+///
+/// The purpose of this data type is to defer conversion as long as possible,
+/// so that no heap allocation is required in case no data is writable.
+pub trait BytesSource {
+    // Returns the total amount of bytes to write
+    fn len(&self) -> usize;
+    /// Limits the source to the new size
+    fn limit(&mut self, new_len: usize);
+    /// The amount of bytes consumed from this source
+    fn bytes_consumed(&self) -> usize;
+    /// The amount of chunks consumed from this source
+    fn chunks_consumed(&self) -> usize;
+    /// Whether a chunk was partially consumed from this source
+    fn partial_chunk_consumed(&self) -> bool;
+    /// Returns the next chunk from the source of owned chunks.
+    ///
+    /// This method will consume parts of the source.
+    // Calling it will yield `Bytes` elements up to the configured `limit`.
+    fn pop_chunk(&mut self) -> Option<Bytes>;
+}
+
+/// A [`BytesSource`] implementation for `&'a mut [Bytes]`.
+///
+/// The type allows to dequeue [`Bytes`] chunks from an array of chunks, up to
+/// a configured limit.
+pub struct BytesArray<'a> {
+    chunks: &'a mut [Bytes],
+    len: usize,
```

Comment:
Is `len` in bytes or in chunks? Looks like the former, but would be good to clarify either by changing the name or adding a comment.

Reply:
It's in bytes. What kind of alternative name would you imagine? `num_bytes`? `bytes_len`?

Reply:
I guess `len` is fine, just add a comment to make it explicit.